ads/google/a4a/performance.js

```diff
@@ -0,0 +1,175 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {EXPERIMENT_ATTRIBUTE} from './traffic-experiments';
+import {dev} from '../../../src/log';
+
+/**
+ * This module provides a fairly crude form of performance monitoring (or
+ * profiling) for A4A code.  It generates individual pings back to Google
+ * servers at key points in the A4A lifecycle and at a few points in the 3p
+ * amp-ad lifecycle, for baseline.
+ *
+ * This is intended to be a short-term solution, for a rough-and-ready form
+ * of profiling.  In particular, it doesn't use high-resolution timers (when
+ * they're available) and it doesn't queue pings for network efficiency.  A
+ * better long-term solution is to integrate `src/performance.js` with
+ * `amp-analytics`.  However, we need a short-term solution quickly.  This
+ * module should go away once we have verified that A4A is performing as
+ * desired.
+ */
+
+/**
+ * Header name for per-ad-slot QQid.
+ * @type {string}
+ */
+export const QQID_HEADER = 'X-QQID';
+
+/** @private */
+const PINGBACK_ADDRESS = 'https://csi.gstatic.com/csi';
+
+/** @private */
+const LIFECYCLE_STAGES = {
+  // Note: Use strings as values here, rather than numbers, so that "0" does
+  // not test as `false` later.
+  adSlotBuilt: '0',
+  urlBuilt: '1',
+  adRequestStart: '2',
+  adRequestEnd: '3',
+  extractCreativeAndSignature: '4',
+  adResponseValidateStart: '5',
+  renderFriendlyStart: '6',
```

Comment:
Because the ad request is the same -- the difference is only in the A4A path, after the creative is received and validated to be / not to be AMP.  Other stages of the pathway are common.